### ISSUE TITLE
D.js has no option to non populate the global namespace

### ISSUE THREAD

**Issue by augusto-altman** [malko/D.js#14]
I couldn't find a way to tell D to not populate the browser's global namespace (window) with the D function. I was looking for something similar to [the jQuery noConflict feature](https://api.jquery.com/jquery.noconflict/).

**Comment by malko** [comment]
jQuery still pollute the global scope when using _noConflict_, it always expose a _jQuery_ and the _$_ variables.
The only thing the _noConflict_ method do is to restore the previous value of _$_ variable still polluting the global namespace with _jQuery_.

So my questions are:
- Do you really need a way to restore the previous _D_ value after loading ?
- _D_ has no other name such as _jQuery_ so how do you expect to access it without being in the global namespace.

Thanks for your interest in D and your responses.

**Comment by augusto-altman** [comment]
That's exactly what I meant when I said I need an option for not pollute the global namespace with D. I'm making a library and I want to avoid possible namespace conflicts with the code that use it. Obviously I expressed myself wrong and luckily you got the idea :). Nevertheless, and just for clarifying things (and for the sake of my honor :P haha), the _noConflict_ function doesn't expose always the _$_ and _jQuery_ values. In fact, if no _$_ and _jQuery_ variables was defined previously to the jQuery load, and you use the _noConflict_ function then you will still not having those variables in the global namespace. And here is why, if you check the [jQuery repo](https://github.com/jquery/jquery/blob/master/src/exports/global.js) you will notice that jQuery always saves the values of _window.$_ and _window.jQuery_ in an internal cache before exposing the jQuery function to the global namepsace. It saves them __no matter if they are defined or not__. So when you use the _noConflict_ function you will restore those values from the cache always. If their were _undefined_ before, they will still being _undefined_ after calling _noConflict_. From a high level point of view, when you use the _noConflict_ function you "clean" the jQuery pollution from the global namespace, it's like jQuery doesn't even exists.

Regarding your concerns, let's consider again the jQuery case. I've prepared an [example in jsFiddle] (http://jsfiddle.net/augustoaltman/5febva20/) using jQuery _noConflict_ and trying to let it as much closer I could to a real scenario. You can see there (the "outputs" will appear in the browser console, so remember to open it :P). that there is no polluted _$_ and _jQuery_ value in the global namespace, but from inside my library I can still use jQuery as usual. 

So basically there is no need to expose a D alias to the global namespace, you could implement something like the [jQuery noConflict code](https://github.com/jquery/jquery/blob/master/src/exports/global.js) and you will get the same effect.

On the other hand, there would be nice for D that it provides AMD support as well. This also avoids the global namespace pollution for those who use RequireJS.

Thank you for the quick response.

Tito (or Augusto, whatever you like the most :P).-

**Comment by malko** [comment]
Ok, can you review the version in branch https://github.com/malko/D.js/tree/14-better-browser-export and tell me if this is ok for you in which case i'll merge it to master asap.
I choosed to use named amd definition as we use almond at work for some project and it require modules to be named to work properly.
cheers

**Comment by augusto-altman** [comment]
The code looks good and it works nice also :). The only little enhancement you could do is to include this conditional line _typeof module !== undefStr && module.exports && (module.exports = defer);_ to the final _if_. Otherwise it would be a tiny, tiny chance to execute the code for expose D in more than one environment what could lead to ugly bugs. Nevertheless, from my side the code is good enough as is, so it's on you to implement or not the little enhancement.

Thank you very much, Tito.

**Comment by malko** [comment]
Thanks tito the new version is available for download through npm and now bower ;)

**Comment by augusto-altman** [comment]
You are awesome! you've gained a new github follower haha
